### PR TITLE
Issue 452/new csproj format and nugets.

### PR DESCRIPTION
**Pull request by Bimble**
#### Description
This is the new csproj format of VS2017, I added them next to the current csproj files and made a seperate solution for it, this to allow VS2015 use. A minor step extra was the nuget generation, which, due to the new format, was easy to realize.

#### Related Issue
[link](https://github.com/QuantConnect/Lean/issues/1686)

#### Motivation and Context
Allows full use of the new changes in Visual Studio, in addition this is also a required step to use the .net standard framework. The nuget part allows easy reuse of the project.

#### Requires Documentation Change
Currently each build triggers a nuget update, this probably should be limited to a release tag/branch or something. This is something that has to be decided upon by the project.
The url to push the nugets to has to probably be changed too.

#### How Has This Been Tested?
Updated the travis.yml to use the new projects, it fully builds and pushes nugets to myget.

#### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Non-functional change (xml comments/documentation/ect)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description` or `feature-<issue#>-<description>`

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -2,12 +2,13 @@ sudo: required
...
 script:
-  - xbuild /p:Configuration=Release /p:VbcToolExe=vbnc.exe QuantConnect.Lean.sln
+  - msbuild /p:Version="2.4.0.1" /t:restore,build /p:Configuration=Release QuantConnect.Lean.VS17.sln
   - mono ./testrunner/NUnit.Runners.2.6.4/tools/nunit-console.exe ./Tests/bin/Release/QuantConnect.Tests.dll --exclude:TravisExclude --labels
 after_success:
-  - coveralls
\ No newline at end of file
+  - coveralls  
+  - curl -O https://dist.nuget.org/win-x86-commandline/latest/nuget.exe
+  - find . -name *.nupkg -not -path "*/packages/*" -not -path "*/testrunner/*"  -exec mono ./nuget.exe push {} $mygetApiKey  -Source https://www.myget.org/F/leantest/api/v2/package \;
```
Comment (jaredbroad) [inline .travis.yml]: This is really cool and we're going to borrow that command line :D But we've got a planned nuget system @Jay-Jay-D is building now that ties into our "build numbers". In the cloud you'll see build numbers in the console -- we're hoping to tie them all together. Would you mind removing this bit?

Comment (jaredbroad) [reply .travis.yml]: image the 3471 is the build number on the end.

File: Tests/QuantConnect.Tests.VS17.csproj [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+    <AssemblyName>QuantConnect.Tests</AssemblyName>
+    <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
+    <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
+  </PropertyGroup>
+  <ItemGroup>
+    <Compile Remove="RegressionAlgorithms\Test_Cash.cs" />
+    <Compile Remove="RegressionAlgorithms\Test_LiveAlgorithm.cs" />
+    <Compile Remove="RegressionAlgorithms\Test_MixedAssets.cs" />
+    <Compile Remove="RegressionAlgorithms\Test_OnEndOfDay.cs" />
+    <Compile Remove="RegressionAlgorithms\Test_SetHolding.cs" />
```
Comment (jaredbroad) [inline Tests/QuantConnect.Tests.VS17.csproj]: Interesting! I was wondering about how it ignores the excluded files.

File: .travis.yml (hunk 1 above)
Comment (Bimble) [reply .travis.yml]: Sure, I can remove it, I take that the nuget generation should also be disabled? As currently there is a flag in the csproj files that generates the nuget.

The planned nuget system is a lot different? I couldn't find that line in the build log, though if you want the build number of travis behind the version in the nuget then the above line of /p:Version="2.4.0.1" could be changed to something like /p:Version="2.4.0.1.$((TRAVIS_JOB_ID - 1))", in addition a condition could be added that it only pushes for the master branch.

Comment (jaredbroad) [reply .travis.yml]: Thanks yes please remove that flag too. We're ready to go with the automated publishing just need your assistance transferring the unofficial packages over.

**Comment by mchandschuh** [comment]
@jaredbroad  -- I'm hesitant to merge this as we'll now have three solutions -- the main one, a VS15 and a VS17 solution. Maybe there's value in replacing `QuantConnect.Lean.sln` with this new `QuantConnect.Lean.VS17.sln` -- then we have our main solution file which is VS17 and there's an alternate solution file for VS15 -- although, we'll still be in the situation where any added files to one will need to be added to the other, and I'm not sure of a good way to enforce this. Do you think it's time we drop VS15 support? VS17 community edition is free and works well. Managing multiple solution files is just somewhat terrifying to me :)

**Comment by mchandschuh** [comment]
#### TO-DO LIST
- [ ] Confirm new project format builds on the build server
- [ ] Consider applying the same transformation to the non-open-sourced projects

**Comment by mchandschuh** [comment]
@Bimble - for the purposes of a status update - I'll be working on testing these changes (and inevitably fixing things that break) in our back-end CI processes, probably later this afternoon or tomorrow morning.

**Comment by mchandschuh** [comment]
Apologies for the long silence on this one -- I'll be working on our CI process tomorrow and evaluating the risk of incorporating these changes. More information to follow tomorrow.

**Comment by mchandschuh** [comment]
Merging this into `feature-452-csproj-format` for now. There's still some work needed in our ci/cd processes to support this change.

**Pull request merged**